v2/permissions/cache.go

```diff
@@ -0,0 +1,125 @@
+package permissions
+
+import (
+	"context"
+	"github.com/ONSdigital/log.go/v2/log"
+	"sync"
+	"time"
+)
+
+// Compiler check to ensure CachingStore implements the Store interface.
+var _ Store = (*CachingStore)(nil)
+
+// CachingStore is a permissions store implementation that caches permission data in memory.
+type CachingStore struct {
+	underlyingStore     Store
+	cachedBundle        *Bundle
+	closing             chan struct{}
+	expiryCheckerClosed chan struct{}
+	cacheUpdaterClosed  chan struct{}
+	lastUpdated         time.Time
+	mutex               sync.Mutex
+}
+
+// NewCachingStore contructs a new instance of CachingStore
+func NewCachingStore(underlyingStore Store) *CachingStore {
+	return &CachingStore{
+		underlyingStore:     underlyingStore,
+		closing:             make(chan struct{}),
+		expiryCheckerClosed: make(chan struct{}),
+		cacheUpdaterClosed:  make(chan struct{}),
+	}
+}
+
+// GetPermissionsBundle returns the cached permission data, or an error if it's not cached.
+func (c *CachingStore) GetPermissionsBundle(ctx context.Context) (*Bundle, error) {
+	if c.cachedBundle == nil {
+		return nil, ErrNotCached
+	}
+
+	return c.cachedBundle, nil
+}
+
+// Update the permissions cache data, by calling the underlying permissions store
+func (c *CachingStore) Update(ctx context.Context) (*Bundle, error) {
+	bundle, err := c.underlyingStore.GetPermissionsBundle(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	c.mutex.Lock()
+	c.cachedBundle = bundle
+	c.lastUpdated = time.Now()
+	c.mutex.Unlock()
+
+	return bundle, nil
+}
+
+// StartExpiryChecker starts a goroutine to continually check for expired cache data.
+//  - checkInterval - how often to check for expired cache data.
+//  - maxCacheTime - how long to cache permissions data before it's expired.
+func (c *CachingStore) StartExpiryChecker(ctx context.Context, checkInterval, maxCacheTime time.Duration) {
+	go func() {
+		defer close(c.expiryCheckerClosed)
+		ticker := time.NewTicker(checkInterval)
+
+		for {
+			select {
+			case <-ticker.C:
+				c.CheckCacheExpiry(ctx, maxCacheTime)
+			case <-c.closing:
+				ticker.Stop()
```

Comment:
I think you might also need to close `ticker.C`. https://pkg.go.dev/time#Ticker.Stop

Reply:
I think this is by design that the ticker channel is not closed https://github.com/golang/go/issues/2650#issuecomment-66061180

Having the ticker stop and function return on `c.closing` should allow it to be fully garbage collected? I couldn't see any examples where it suggested the channel should be closed.

Reply:
Sorry I misread the godoc - They are saying that stop stops the ticker but if you care about go routines still accessing it while its stopping then you need to close the channel - at least I think thats what they mean.

For this that doesn't appear to be a problem so resolving comment.